File: Chess/Chess.GameLogic/Detectors/AvailableCagesHelper.cs
using Chess.GameLogic.Extensions;
using Chess.GameLogic.Models;

namespace Chess.GameLogic.Detectors
{
    internal class AvailableCagesHelper
    {
        public List<PiecePositionDto> GetAvailableCagesMovingWith(IEnumerable<PieceDto> pieces, PieceDto current, int vertPosDiffKoef, int horPosDiffKoef)
        {
            var vertPos = current.Position.PosY + vertPosDiffKoef;
            var horPos = current.Position.PosX + horPosDiffKoef;
            var availableCages = new List<PiecePositionDto>();

            while (true)
            {
                var checkedPos = new PiecePositionDto(vertPos, horPos);
                var canBeSetedToPos = pieces.CanBeSetedToPosition(checkedPos, current.Color, out bool isEnemyOnCage);
                if (canBeSetedToPos)
                    availableCages.Add(checkedPos);

                if (!canBeSetedToPos || isEnemyOnCage)
                    return availableCages;

                vertPos += vertPosDiffKoef;
                horPos += horPosDiffKoef;
            }
        }
    }
}

File: Front/chess/src/modules/ChessGame/helpers/AvailableCagesHelper.js
import { PiecesHelper } from "./PiecesHelper"

export class AvailableCagesHelper{
    static getAvailableCagesMovingWith(pieces, piece, vertDiffKoef, horDiffKoef){
        let posX = piece.validPosition.posX + horDiffKoef
        let posY = piece.validPosition.posY + vertDiffKoef
        const availableCages = []
        while(true){
            const checkedPos = {posX: posX, posY: posY}
            const canBeSetedToPositionResult = PiecesHelper.canBeSetedToPosition(pieces, checkedPos, piece.color)
            if(canBeSetedToPositionResult.canBeSetted)
                availableCages.push(checkedPos)

            if(!canBeSetedToPositionResult.canBeSetted || canBeSetedToPositionResult.isEnemyOnCage)
                return availableCages

            posX += horDiffKoef
            posY += vertDiffKoef
        }
    }
}